<launch>
  <arg name="uav_name" default="$(optenv UAV_NAME uav38)"/>

  <arg name="standalone" default="false"/>

  <arg unless="$(arg standalone)" name="nodelet" value="load"/>
  <arg     if="$(arg standalone)" name="nodelet" value="standalone"/>
  <arg unless="$(arg standalone)" name="nodelet_manager" value="$(arg uav_name)_uvdar_nodelet_manager"/>
  <arg     if="$(arg standalone)" name="nodelet_manager" value=""/>

  <arg name="threshold" default="100"/>

  <arg name="calibrations_folder" default="$(find mrs_uav_deployment)/config/uvdar_calibrations"/>
  
  <param name="use_sim_time" value="true"/>



  <!-- Node Settings -->
  <arg name="output" default="screen"/>
  <arg name="proc" default="false"/>
  <arg name="view" default="false"/>
  <arg name="calib" default="false"/>

  <arg name="profiling" default="false"/>
  <arg name="debug" default="false"/>
  <arg name="visual_debug" default="false"/>
  <arg name="publish_visualization" default="false"/>
  <arg name="visualization_rate" default="15"/>

  <arg name="gui" default="true"/>
  <arg name="publish" default="true"/>
  
  <arg name="id1" value="0"/>
  <arg name="id2" value="1"/>
  <arg name="id3" value="2"/>
  <arg name="id4" value="3"/>
  <arg name="id5" value="4"/>
  <arg name="id6" value="5"/>
  <arg name="id7" value="6"/>
  <arg name="id8" value="7"/>
  <arg name="id9" value="8"/>
  <arg name="id10" value="9"/>
  <arg name="id11" value="10"/>
  <arg name="id12" value="11"/>

  <arg name="beacon" default="false"/>

  <arg name="sequence_file" default="$(find uvdar_core)/config/selected.txt"/>

  <arg name="max_px_shift_y" default="3"/>
  <arg name="max_px_shift_x" default="3"/>
  <arg name="max_zeros_consecutive" default="2"/>
  <arg name="max_ones_consecutive" default="2"/>
  <arg name="stored_seq_len_factor" default="3"/>
  <arg name="max_buffer_length" default="1000"/>
  <arg name="poly_order" default="4"/>
  <arg name="decay_factor" default="0.01"/>
  <arg name="confidence_probability" default="95.0"/>
  <arg name="allowed_BER_per_seq" default="0"/>
  <arg name="std_threshold_poly_reg" default="1"/>

  <arg name="use_4DHT" default="false"/>
  <!--4DHT values-->
  <arg name="accumulator_length"         default="18"/>
  <arg name="pitch_steps"         default="16"/>
  <arg name="yaw_steps"         default="16"/>
  <arg name="max_pixel_shift"         default="4"/>


  <!--KF STUFF-->
  <arg name="output_frame"      default="local_origin"/>
  <arg name="use_velocity"      default="false"/>
  <arg name="output_framerate"  default="10"/>

  <arg name="indoor"            default="false"/>
  <arg name="use_two_cameras"   default="true"/>
  <arg name="use_three_cameras"   default="false"/>
  <arg name="anonymous_measurements"   default="false"/>



  <arg name="model_file" default="$(find uvdar_core)/config/models/quadrotor.txt"/>
  <!-- ideal angle: 1.221730476 - pi/2 -->

      <!-- <origin xyz="0.108 0 0.085" rpy="1.57079632679 4.71238898038 1.57079632679" /> -->
  <!--Usage: static_transform_publisher x y z yaw pitch roll frame_id child_frame_id  period(milliseconds)-->
<!-- OR -->
  <group ns="$(arg uav_name)">

    <node pkg="nodelet" type="nodelet" name="$(arg uav_name)_uvdar_nodelet_manager" args="manager" output="screen" >
    <!-- <node pkg="nodelet" type="nodelet" name="$(arg uav_name)_uvdar_nodelet_manager" args="manager" output="screen" launch-prefix="debug_roslaunch"> -->
      <param name="num_worker_threads" value="8" />
    </node>

    <!-- <node name="bluefox_emulator" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARBluefoxEmulator $(arg nodelet_manager)" output="screen" respawn="false" launch-prefix="bash -c 'sleep 5; $0 $@'">
      <rosparam param="camera_output_topics"> ["camera_back"] </rosparam>
      <rosparam param="calib_files"> ["default"] </rosparam>
      <remap from="~camera_back" to="/$(arg uav_name)/uvdar_bluefox_back/image_raw"/>
      <remap from="~camera_back_transfer" to="/gazebo/$(arg uav_name)/uvdar_bluefox_back/image_raw"/>
    </node> -->
      
    
    <node name="blink_processor" pkg="nodelet" type="nodelet" args="load uvdar/UVDARBlinkProcessor $(arg nodelet_manager)" output="screen" respawn="false">
    <!-- <node name="UVDARBlinkProcessor" pkg="nodelet" type="nodelet" args="load uvdar/UVDARBlinkProcessor $(arg nodelet_manager)" output="screen" respawn="false" launch-prefix="debug_roslaunch"> -->

      <param name="uav_name" type = "string" value="$(arg uav_name)"/>
      <param name="debug"		type="bool" value="$(arg debug)"/> 
      <param name="gui"             type="bool" value="$(arg gui)"/>
      
      <!-- Changeable for each nodelet independently. - no arguments! -->
      <param name="publish_visualization"             type="bool" value="true"/>
      <param name="use_camera_for_visualization"         type="bool" value="false"/>
      <param name="visualization_rate"         type="double" value="2"/>

      <param name="use_4DHT"             type="bool" value="$(arg use_4DHT)"/>
    

      <param name="max_px_shift_x"          type="int"    value="$(arg max_px_shift_x)"/>
      <param name="max_px_shift_y"          type="int"    value="$(arg max_px_shift_y)"/>
      <param name="max_zeros_consecutive"   type="int"    value="$(arg max_zeros_consecutive)"/>
      <param name="max_ones_consecutive"    type="int"    value="$(arg max_ones_consecutive)"/>
      <param name="stored_seq_len_factor"   type="int"    value="$(arg stored_seq_len_factor)"/>  
      <param name="max_buffer_length"       type="int"    value="$(arg max_buffer_length)"/>  
      <param name="poly_order"              type="int"    value="$(arg poly_order)"/>
      <param name="decay_factor"            type="double" value="$(arg decay_factor)"/>    
      <param name="confidence_probability"  type="double" value="$(arg confidence_probability)"/>
      <param name="allowed_BER_per_seq"     type="int"    value="$(arg allowed_BER_per_seq)"/>
      <param name="std_threshold_poly_reg"  type="double" value="$(arg std_threshold_poly_reg)"/>
      
      <param name="accumulator_length"         type="int" value="$(arg accumulator_length)"/>
      <param name="pitch_steps"         type="int" value="$(arg pitch_steps)"/>
      <param name="yaw_steps"         type="int" value="$(arg yaw_steps)"/>
      <param name="max_pixel_shift"         type="int" value="$(arg max_pixel_shift)"/>
      <param name="nullify_radius"         type="int" value="5"/>

 
      <param name="sequence_file"         type="string" value="$(arg sequence_file)"/>

      <rosparam param="camera_topics"> ["camera_left", "camera_right"] </rosparam>
      <rosparam param="points_seen_topics"> ["points_seen_left", "points_seen_right"] </rosparam>
      <rosparam param="blinkers_seen_topics"> ["blinkers_seen_left", "blinkers_seen_right"] </rosparam>
      <rosparam param="estimated_framerate_topics"> ["estimated_framerate_left", "estimated_framerate_right"] </rosparam>
      <rosparam param="omta_logging_topics"> ["omta_logging_left", "omta_logging_right"] </rosparam>
      <rosparam param="omta_all_seq_info_topics"> ["omta_all_seq_info_left","omta_all_seq_info_right"] </rosparam>


      <remap from="~camera_left" to="/$(arg uav_name)/uvdar_bluefox/left/image_raw_camp"/>
      <remap from="~camera_right" to="/$(arg uav_name)/uvdar_bluefox/right/image_raw_camp"/>
      <remap from="~points_seen_left" to="/$(arg uav_name)/uvdar/points_seen_left"/>
      <remap from="~points_seen_right" to="/$(arg uav_name)/uvdar/points_seen_right"/>
      <remap from="~blinkers_seen_left" to="/$(arg uav_name)/uvdar/blinkers_seen_left_camp"/>
      <remap from="~blinkers_seen_right" to="/$(arg uav_name)/uvdar/blinkers_seen_right_camp"/>
      <remap from="~estimated_framerate_left" to="/$(arg uav_name)/uvdar/estimated_framerate_left_camp"/>
      <remap from="~estimated_framerate_right" to="/$(arg uav_name)/uvdar/estimated_framerate_right_camp"/>

      <remap from="~visualization" to="/$(arg uav_name)/uvdar/blink_visualization/image_raw_camp"/> <!-- TODO: not publishing video if not commented in-->
      <remap from="~omta_logging_left" to="/$(arg uav_name)/uvdar/omta_logging_left_camp"/>
      <remap from="~omta_logging_right" to="/$(arg uav_name)/uvdar/omta_logging_right_camp"/>
      <remap from="~omta_all_seq_info_left" to="/$(arg uav_name)/uvdar/omta_all_seq_info_left_camp"/>
      <remap from="~omta_all_seq_info_right" to="/$(arg uav_name)/uvdar/omta_all_seq_info_right_camp"/>
      <remap from="~omta_all_seq_info_back" to="/$(arg uav_name)/uvdar/omta_all_seq_info_back"/>

    </node>
    





      <!-- <remap from="~blinkers_seen_left_camp" to="/$(arg uav_name)/uvdar/blinkers_seen_left_camp"/> -->
      <!-- <remap from="~blinkers_seen_right_camp" to="/$(arg uav_name)/uvdar/blinkers_seen_right_camp"/> -->
      <!-- <remap from="~estimated_framerate_left_camp" to="/$(arg uav_name)/uvdar/estimated_framerate_left_camp"/> -->
      <!-- <remap from="~estimated_framerate_right_camp" to="/$(arg uav_name)/uvdar/estimated_framerate_right_camp"/> -->

      <!-- <rosparam param="calib_files" subst_value="True"> ["default", "default"] </rosparam> -->

      <!-- <remap from="~constituentPoses1" to="/$(arg uav_name)/uvdar/constituentPosesL_camp"/> -->
      <!-- <remap from="~constituentPoses2" to="/$(arg uav_name)/uvdar/constituentPosesR_camp"/> -->

      <!-- <remap from="~measuredPoses1" to="/$(arg uav_name)/uvdar/measuredPosesL_camp"/> -->
      <!-- <remap from="~measuredPoses2" to="/$(arg uav_name)/uvdar/measuredPosesR_camp"/> -->
      
    <!-- </node> -->




  
    <!-- <node pkg="uvdar_core" type="uvdar_filter_node" name="filterUvdar" output="screen" launch-prefix="debug_roslaunch"> -->
      <!-- <node pkg="uvdar_core" type="uvdar_filter_node" name="filterUvdar" output="screen"> -->
        <!-- <param name="debug"             value="$(arg debug)" /> -->
        <!-- <param name="uav_name"          value="$(arg uav_name)" /> -->
        <!-- <param name="output_frame"      value="$(arg output_frame)"/> -->
        <!-- <param name="output_framerate"  value="$(arg output_framerate)" /> -->

        <!-- <param name="indoor"            value="$(arg indoor)" /> -->
        <!-- <param name="use_velocity"      value="$(arg use_velocity)" /> -->
        <!-- <param name="anonymous_measurements"      value="$(arg anonymous_measurements)" /> -->

        <!-- <rosparam     unless="$(arg use_two_cameras)" param="measured_poses_topics"> ["measured_poses"] </rosparam> -->
        <!-- <remap        unless="$(arg use_two_cameras)" from="~measured_poses" to="/$(arg uav_name)/uvdar/measuredPoses"/> -->

        <!-- <rosparam     if="$(arg use_two_cameras)" param="measured_poses_topics"> ["measured_poses_L", "measured_poses_R"] </rosparam> -->
        <!-- <remap        if="$(arg use_two_cameras)" from="~measured_poses_L" to="/$(arg uav_name)/uvdar/measuredPosesL"/> -->
        <!-- <remap        if="$(arg use_two_cameras)" from="~measured_poses_R" to="/$(arg uav_name)/uvdar/measuredPosesR"/> -->

        <!-- <rosparam     if="$(arg use_three_cameras)" param="measured_poses_topics"> ["measured_poses_L", "measured_poses_R", "measured_poses_B"] </rosparam> -->
        <!-- <remap        if="$(arg use_three_cameras)" from="~measured_poses_L" to="/$(arg uav_name)/uvdar/measuredPosesL"/> -->
        <!-- <remap        if="$(arg use_three_cameras)" from="~measured_poses_R" to="/$(arg uav_name)/uvdar/measuredPosesR"/> -->
        <!-- <remap        if="$(arg use_three_cameras)" from="~measured_poses_B" to="/$(arg uav_name)/uvdar/measuredPosesB"/> -->


        <!-- <remap from="~filtered_poses" to="/$(arg uav_name)/uvdar/filteredPoses"/>
        <remap from="~filtered_poses/tentative" to="/$(arg uav_name)/uvdar/filteredPoses/tentative"/> -->

    <!-- </node> -->

  </group>


</launch>
